beacon_node/beacon_chain/src/data_availability_checker.rs

```diff
@@ -511,6 +525,92 @@ impl<T: BeaconChainTypes> DataAvailabilityChecker<T> {
             block_cache_size: self.availability_cache.block_cache_size(),
         }
     }
+
+    pub fn reconstruct_data_columns(
+        &self,
+        block_root: &Hash256,
+    ) -> Result<DataColumnReconstructionResult<T::EthSpec>, AvailabilityCheckError> {
+        let pending_components = match self
+            .availability_cache
+            .check_and_set_reconstruction_started(block_root)
+        {
+            ReconstructColumnsDecision::Yes(pending_components) => pending_components,
+            ReconstructColumnsDecision::No(reason) => {
+                return Ok(DataColumnReconstructionResult::NotRequired(reason));
```

Comment:
I wanted to point out that the Error variant here is called `NotRequired` but `ReconstructionDecision` includes "not enough columns" which (unlike every other variant) doesn't actually mean it's not required, only that it can't be done yet. I thought you might instead change things to:
```
#[derive(Debug)]
pub enum RejectionCriteria {
    BlockAlreadyImported,
    AlreadyStarted,
    NotRequiredForFullNode,
    AllColumnsReceived,
    NotEnoughColumns,
}

pub enum DataColumnReconstructionResult<E: EthSpec> {
    Success(AvailabilityAndReconstructedColumns<E>),
    NotRequired(RejectionCriteria),
    Pending(RejectionCriteria),
}

pub(crate) enum ReconstructColumnsDecision<E: EthSpec> {
    Yes(PendingComponents<E>),
    No(RejectionCriteria),
}

impl<E: EthSpec> From<RejectionCriteria> for DataColumnReconstructionResult<E> {
    fn from(criteria: RejectionCriteria) -> Self {
        match criteria {
            RejectionCriteria::NotEnoughColumns => DataColumnReconstructionResult::Pending(criteria),
            _ => DataColumnReconstructionResult::NotRequired(criteria),
        }
    }
}
```

Reply:
Ah yeah I thought about having an enum for all the reasons, and I decided to go with string instead because we don't really need to handle each variant differently, the main usage of the `reason` is for metric label, and it requires a string, hence I went with it for simplicity and efficiency:

https://github.com/sigp/lighthouse/blob/0e6eaa227890194983cd523d5e6be498907dfdad/beacon_node/beacon_chain/src/beacon_chain.rs#L3199-L3202

> the Error variant here is called NotRequired but ReconstructionDecision includes "not enough columns" which (unlike every other variant) doesn't actually mean it's not required, only that it can't be done yet.

Yes that's right, if this is confusing I could rename `DataColumnReconstructionResult::NotRequired` to something else?

Reply:
I've renamed `NotRequired` to `NotStarted`.